### ISSUE TITLE
Bugs: cannot provide/override props of SpinButton.shinyInput 

### ISSUE THREAD

**Issue by jonekeat** [Appsilon/shiny.fluent#215]
### Code example

```r
library(shiny.fluent)

shinyApp(
  ui = div(
    tags$script(HTML(
      "
      /*
        defined salary min, step & default
      */
      const xyz = {
        salaryMin: 1000,
        salaryStep: 100,
        salaryDefault: 10000
      };
      
      /*
        format string or number to currency format,
        eg. '123456.789' to '123,456.79'
      */
      const formatCurrency = (value, minimumFractionDigits = 2, maximumFractionDigits = 2) => {
        const currency = value ? (new Intl.NumberFormat('en-US', {
          minimumFractionDigits: minimumFractionDigits,
          maximumFractionDigits: maximumFractionDigits
        }).format(value)) : undefined;
        return ((currency && currency != \"NaN\") ? currency : undefined);
      };
      
      /*
        format currency to number,
        eg. '123,456.78' to '123456.78'
      */
      const currencyToNumber = (value) => {
        const numericValue = value ? Number(value.replace(/,/g, '')) : undefined;
        return (numericValue ? numericValue : undefined);
      };
      
      /*
        increment the value of `xyz-salary` (or return nothing to keep the previous value if invalid)
        no maximum limit, step: salaryStep
      */
      xyz['salaryOnIncrement'] = function salaryOnIncrement(value, event) {
        const numericValue = currencyToNumber(value);
        if (numericValue !== undefined) {
          const currency = formatCurrency(numericValue + xyz.salaryStep);
          if (currency !== undefined) {
            return String(currency);
          }
        }
      };
      
      /*
        decrement the value of `xyz-salary` (or return nothing to keep the previous value if invalid)
        minimum limit: salaryMin, step: salaryStep
      */
      xyz['salaryOnDecrement'] = function salaryOnDecrement(value, event) {
        const numericValue = currencyToNumber(value);
        if (numericValue !== undefined) {
          const currency = formatCurrency(Math.max(numericValue - xyz.salaryStep, xyz.salaryMin));
          if (currency !== undefined) {
            return String(currency);
          }
        }
      };
      
      /*
        clamp the value of `xyz-salary` within the valid range (or return nothing to keep the previous value
        if there's not valid numeric input)
        no maximum, minimum limit: salaryMin
      */
      xyz['salaryOnValidate'] = function salaryOnValidate(value, event) {
        const numericValue = currencyToNumber(value);
        if (numericValue !== undefined) {
          const currency = formatCurrency(Math.max(numericValue, xyz.salaryMin));
          if (currency !== undefined) {
            return String(currency);
          }
        }
      };
      
      "
    )),
    SpinButton.shinyInput(
      inputId = "salary",
      className = "employee-salary",
      `data-type` = 'currency',
      value = JS("String(formatCurrency(xyz.salaryDefault))"),
      min = JS("xyz.salaryMin"),
      onValidate = JS("xyz.salaryOnValidate"),
      onIncrement = JS("xyz.salaryOnIncrement"),
      onDecrement = JS("xyz.salaryOnDecrement"),
      incrementButtonAriaLabel = "Increase value by 100",
      decrementButtonAriaLabel = "Decrease value by 100"
    ),
    br(),
    SpinButton(
      className = "employee-salary2",
      `data-type` = 'currency',
      defaultValue = JS("String(formatCurrency(xyz.salaryDefault))"),
      min = JS("xyz.salaryMin"),
      onValidate = JS("xyz.salaryOnValidate"),
      onIncrement = JS("xyz.salaryOnIncrement"),
      onDecrement = JS("xyz.salaryOnDecrement"),
      onChange = JS("(event, value) => Shiny.setInputValue('salary2', currencyToNumber(value))"), # input must prefix with ns, see \inst\app\js\utils.js for `currencyToNumber()`
      incrementButtonAriaLabel = "Increase value by 100",
      decrementButtonAriaLabel = "Decrease value by 100"
    ),
    textOutput("spinValue"),
    textOutput("spinValue2"),
    ActionButton.shinyInput("button", text = "Action Button")
  ),
  server = function(input, output) {
    output$spinValue <- renderText({
      sprintf("Value: %s", input$salary)
    })
    
    output$spinValue2 <- renderText({
      sprintf("Value: %s", input$salary2)
    })
    
    observeEvent(input$button, {
      updateSpinButton.shinyInput(
        inputId = "salary",
        disable = TRUE,
        value = 0
      )
    }) 
    
  }
)

```

### Bug description

Hi thanks again for this amazing package!

I discover 2 bugs here:
1. `SpinButton.shinyInput` cannot mimic the behavior of `SpinButton` even we provide same the props, I wonder if props in `SpinButton` are correctly passed to `SpinButton.shinyInput` or not
2. `updateSpinButton.shinyInput` failed to disable `SpinButton.shinyInput` from server side

Please let me know if you need more info

### Expected behavior

`SpinButton.shinyInput` should be able to reproduce the `SpinButton` example, and `updateSpinButton.shinyInput` should be able to update the `SpinButton.shinyInput` to disable

**Comment by kamilzyla** [comment]
Thanks @jonekeat for your report and for your enthusiasm about shiny.fluent! :smiley:

### Plain components vs `.shinyInput` adapters

All the components from the underlying Fluent UI library are available as "plain components" which directly expose their React interface. This works great for many cases, but the differences between React and Shiny often make it inconvenient to use these plain components for handling user input.

That's why we have `.shinyInput` adapters. They are designed to bring the interface of the selected components closer to what you would expect from vanilla shiny, so e.g. `shiny.fluent::SpinButton.shinyInput` feels a lot like `shiny::numericInput()`. The mapping is not perfect however. We strive to retain maximum flexibility in the adapters, but sometimes it is hard or outright impossible.

Thus, in general you should not expect the `.shinyInput` adapters to always cover the full range of features of the underlying components. For some advanced use cases you'll have to use the React interface of the plain component.

In your particular case, the following differences between `SpinButton` and `SpinButton.shinyInput` are relevant:
* The plain components do not send their initial value to the server, so `input$spinValue2` is initially `NULL`.
* The `SpinButton.shinyInput` expects the `value` to be convertible to a number via `Number(value)`. Unfortunately, this means you cannot use `onIncrement`, `onDecrement` and `onValidate` to provide custom value formatting like in your example. You need to use `SpinButton` for that.

To sum up, the behavior you encountered is a limitation that we have agreed to. However, your report did help me to spot a problem with how the adapter uses the `value` prop - thanks! :smile: I have created issue #216 to track it and included some thoughts inspired by your use case. Perhaps we could support custom value formatting with our adapter?

### Disabling

To disable a component, you need to use the `disabled` prop (not `disable`):

```r
library(shiny)
library(shiny.fluent)

shinyApp(
  ui = tagList(
    SpinButton.shinyInput("spin"),
    Checkbox.shinyInput("disabled", label = "Disabled"),
  ),
  server = function(input, output) {
    observe({
      updateSpinButton.shinyInput(
        inputId = "spin",
        disabled = input$disabled
      )
    })
  }
)

```
